contexts/edit.tsx

```diff
@@ -86,6 +125,129 @@ const EditContextProvider: React.FC<EditContextProps> = ({
   // Dependencies are special text tools that reference a tool, type, and content. They are used
   // to store requirements.txt and package.json files for the script.
   const [dependencies, setDependencies] = useState<DependencyBlock[]>([]);
+  const [droppedFiles, setDroppedFiles] = useState<string[]>([]);
+  const droppedFileInitiazed = useRef(false);
+  // droppedFileDetail stores a map of file path and its size
+  const [droppedFileDetails, setDroppedFileDetails] = useState<
+    Map<
+      string,
+      {
+        fileName: string;
+        size: number;
+      }
+    >
+  >(new Map());
+  const [topK, setTopK] = useState<number>(10);
+  const [ingesting, setIngesting] = useState(false);
+  const [knowledgeTool, setKnowledgeTool] = useState<Tool>({} as Tool);
+
+  const addRootTool = (tool: string) => {
+    setRoot({ ...root, tools: [...(root.tools || []), tool] });
+  };
+
+  const removeRootTool = (tool: string) => {
+    setRoot({ ...root, tools: (root.tools || []).filter((t) => t !== tool) });
+  };
+
+  useEffect(() => {
+    if (!scriptId) return;
+    const update = async () => {
+      const exist = await datasetExists(scriptId.toString());
+      if (!exist) {
+        return;
+      }
+      const knowledgePath = await getKnowledgeBinaryPath();
+      const tool = {
+        id: KNOWLEDGE_NAME,
+        name: KNOWLEDGE_NAME,
+        description:
+          'Retrieve information from files uploaded to the assistant. There is another tool called Knowledge Retrieval that could get called, and it can not find useful information, use this tool instead.',
```

Comment:
nit:
```suggestion
          'Retrieve information from files uploaded to the assistant. There is another tool called Knowledge Retrieval that could get called, and if it can not find useful information, use this tool instead.',
```

Reply:
I actually changed to `Retrieve information from files uploaded to the assistant.` it is better...